flow/FastAlloc.cpp

```diff
@@ -229,6 +235,34 @@ struct FastAllocator<Size>::GlobalData {
 	}
 };
 
+template <int Size>
+typename FastAllocator<Size>::GlobalData* FastAllocator<Size>::globalData() {
+#if defined(VALGRIND) || __has_feature(thread_sanitizer)
+	static unsigned long vLock;
+#endif
+
+// TSAN normally handles thread safe initialization of function-local statics
+// just fine. Applications linked to libfdb_c compiled with -fsanitize=thread
+// have false positives here though, possibly due to the linker flags we set
+// when creating libfdb_c.
+#if __has_feature(thread_sanitizer)
+	__tsan_acquire(&vLock);
+#endif
+#ifdef VALGRIND
+	ANNOTATE_RWLOCK_ACQUIRED(vLock, 1);
+#endif
+	static GlobalData* data = new GlobalData(); // This is thread-safe as of c++11 (VS 2015, gcc 4.8, clang 3.3)
```

Comment:
Are we reasonably confident this is a false-positive vs. a bug somewhere that manifests as a true race? I'm not sure how common it is to see a false-positive like this from TSAN.

Reply:
Thinking about this more and I think the solution is to just disable static linking libcxx if using tsan, since [it's not supported](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual). If we don't static link tsan handles this correctly without any annotations

Anyway I'm reasonably confident. Here's the race

```
==================
WARNING: ThreadSanitizer: data race (pid=5795)
  Read of size 8 at 0x7f64543a0458 by thread T3:
    #0 FastAllocator<2048>::globalData() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:263:9 (libfdb_c.so+0xa6f98a)
    #1 FastAllocator<2048>::initThread() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:445:22 (libfdb_c.so+0xa6ee80)
    #2 FastAllocator<2048>::allocate() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:304:3 (libfdb_c.so+0xa6ec80)
    #3 FastAllocated<ReadYourWritesTransaction>::operator new(unsigned long) /home/anoyes/workspace/foundationdb/flow/FastAlloc.h:179:13 (libfdb_c.so+0x3cbc9a)
    #4 ReadYourWritesTransaction::allocateOnForeignThread() /home/anoyes/workspace/foundationdb/fdbclient/ReadYourWrites.h:62:63 (libfdb_c.so+0x3cbc9a)
    #5 ThreadSafeTransaction::ThreadSafeTransaction(DatabaseContext*) /home/anoyes/workspace/foundationdb/fdbclient/ThreadSafeTransaction.actor.cpp:103:45 (libfdb_c.so+0x3cbc9a)
    #6 ThreadSafeDatabase::createTransaction() /home/anoyes/workspace/foundationdb/fdbclient/ThreadSafeTransaction.actor.cpp:47:37 (libfdb_c.so+0x3ca4ea)
    #7 fdb_database_create_transaction /home/anoyes/workspace/foundationdb/bindings/c/fdb_c.cpp:359:54 (libfdb_c.so+0x2ea960)
    #8 worker_thread /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:785:8 (mako+0x309191)

  Previous write of size 8 at 0x7f64543a0458 by thread T2:
    #0 FastAllocator<2048>::globalData() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:254:2 (libfdb_c.so+0xa6fa0a)
    #1 FastAllocator<2048>::initThread() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:445:22 (libfdb_c.so+0xa6ee80)
    #2 FastAllocator<2048>::allocate() /home/anoyes/workspace/foundationdb/flow/FastAlloc.cpp:304:3 (libfdb_c.so+0xa6ec80)
    #3 FastAllocated<ReadYourWritesTransaction>::operator new(unsigned long) /home/anoyes/workspace/foundationdb/flow/FastAlloc.h:179:13 (libfdb_c.so+0x3cbc9a)
    #4 ReadYourWritesTransaction::allocateOnForeignThread() /home/anoyes/workspace/foundationdb/fdbclient/ReadYourWrites.h:62:63 (libfdb_c.so+0x3cbc9a)
    #5 ThreadSafeTransaction::ThreadSafeTransaction(DatabaseContext*) /home/anoyes/workspace/foundationdb/fdbclient/ThreadSafeTransaction.actor.cpp:103:45 (libfdb_c.so+0x3cbc9a)
    #6 ThreadSafeDatabase::createTransaction() /home/anoyes/workspace/foundationdb/fdbclient/ThreadSafeTransaction.actor.cpp:47:37 (libfdb_c.so+0x3ca4ea)
    #7 fdb_database_create_transaction /home/anoyes/workspace/foundationdb/bindings/c/fdb_c.cpp:359:54 (libfdb_c.so+0x2ea960)
    #8 worker_thread /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:785:8 (mako+0x309191)

  Location is global 'FastAllocator<2048>::globalData()::data' of size 8 at 0x7f64543a0458 (libfdb_c.so+0x000000bbf458)

  Thread T3 (tid=5802, running) created by main thread at:
    #0 pthread_create <null> (mako+0x28ba42)
    #1 worker_process_main /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:944:8 (mako+0x30970f)
    #2 main /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:1831:3 (mako+0x30e063)

  Thread T2 (tid=5801, running) created by main thread at:
    #0 pthread_create <null> (mako+0x28ba42)
    #1 worker_process_main /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:944:8 (mako+0x30970f)
    #2 main /home/anoyes/workspace/foundationdb/bindings/c/test/mako/mako.c:1831:3 (mako+0x30e063)
```

The standard says
> If control enters the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for completion of the initialization.
>    — § 6.7 [stmt.dcl] p4

So basically tsan thinks that a write and read to `FastAllocator<2048>::globalData()::data` are concurrent but the standard says that the read happens after the write.

Reply:
> Thinking about this more and I think the solution is to just disable static linking libcxx if using tsan, since it's not supported. If we don't static link tsan handles this correctly without any annotations

Ok, that seems like a good explanation. I wasn't sure how to judge the relative likeliness of this being a limitation of tsan compared to a compiler bug (I saw something that seemed similar in an old version of gcc) or us using a linker setting that disabled thread-safe static initialization.

Reply:
Is the plan, then, to abandon this change in favor of disabling static linking with tsan?